### PR TITLE
Presence Containers

### PR DESCRIPTION
**Pull request by hansthienpondt**
This commit adds support for YANG presence containers. It adds a tag to
the field of generated Go Struct. When rendering JSON documents, before
deleting/omitting empty containers, it checks for this struct field tag.

As per the OpenConfig style guide, YANG presence containers are not
used, as such this would only be used with 3rd party YANG models. As
such a CLI knob is added to generator.go to enable this behaviour with
-yangpresence=true

### CODE REVIEW

**Comment by hansthienpondt** [comment]
proposed fix for https://github.com/openconfig/ygot/issues/329

File: ygot/render_test.go [hunk 1]
```diff
@@ -2933,6 +2978,69 @@ func TestConstructJSON(t *testing.T) {
 		},
 		wantErr:     true,
 		wantJSONErr: true,
+	}, {
+		name: "device example with presence containers #1",
+		in: &ucExampleDevice{
+			Bgp: &ucExampleBgp{},
+		},
+		wantIETF: map[string]interface{}{"bgp": map[string]interface{}{}},
+		wantSame: true,
```
Comment (wenovus) [inline ygot/render_test.go:2987]: Add a testcase with `Bgp: nil`.

**Review by wenovus — CHANGES_REQUESTED** [review]
This looks pretty good to me. One other suggestion is to add an integration test by copying testdata/modules/openconfig-simple.yang to a new file, adding a presence container, and a new test case (copy over one of the test cases) in codegen_test.go just to make sure it works end-to-end.

File: ygot/render_test.go (hunk 1 above)
Comment (hansthienpondt) [reply ygot/render_test.go:2987]: Ack, Thanks Wen!

File: ygen/gogen.go [hunk 2]
```diff
@@ -1699,6 +1699,14 @@ func writeGoStruct(targetStruct *Directory, goStructElements map[string]*Directo
 
 		metadataTagBuf.WriteString(` ygotAnnotation:"true"`)
 
+		if goOpts.AddYangPresence {
+			// a presence container is an unimplemented keyword in goyang.
+			// if and when this changes, the field lookup below would need to change as well.
```
[the remark this answers is not in the document]
Comment (hansthienpondt) [reply ygen/gogen.go:1704]: Ack, Thanks Wen!

File: util/yang_test.go [hunk 3]
```diff
@@ -1019,6 +1019,33 @@ func TestIsYgotAnnotation(t *testing.T) {
...
+		No  *string
+	}
+
+	tests := []struct {
+		name string
+		in   reflect.StructField
+		want bool
+	}{{
+		name: "yangPresence container/field",
+		in:   reflect.TypeOf(testStruct{}).Field(0),
```
[the remark this answers is not in the document]
Comment (hansthienpondt) [reply util/yang_test.go:1034]: Ack, Thanks Wen!

[the remark this answers is not in the document]
Comment (hansthienpondt) [reply util/yang_test.go:1034]: Hi Wen,  `FieldByName(name string) (StructField, bool)` returns 2 values, so I'd need to assign a variable first...
The test cases for `IsYgotAnnotation()` work similar using `reflect.typeOf(struct{}).Field(int)`, 

Or would you rather have me define:
```		
    in: reflect.StructField{
        Name: "Yes",
        Tag:  `yangPresence:"true"`,
    },
```
and not define a testStruct{} at all?

[the remark this answers is not in the document]
Comment (hansthienpondt) [reply util/yang_test.go]: I went with `FieldByName` anyhow, as this is resembles the real scenario as close as possible. Thnx!

File: ygot/render_test.go [hunk 4]
```diff
@@ -1658,6 +1658,51 @@ type exampleTransportAddressBinary struct {
 
 func (*exampleTransportAddressBinary) IsExampleTransportAddress() {}
 
+// ucExampleDevice and the following structs are a set of structs used for more
+// complex testing in TestConstructIETFJSON, these structs are used to test for
+// YANG presence containers.
+
+type ucExampleDevice struct {
+	Bgp *ucExampleBgp `path:"bgp" yangPresence:"true"`
```
Comment (wenovus) [inline ygot/render_test.go:1666]: I'm also thinking of the cases of a presence container within a non-presence container or a list entry. I think they'll work but a test would guarantee it.

**Review by wenovus — CHANGES_REQUESTED** [review]

File: util/yang_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (wenovus) [reply util/yang_test.go]: Oh i see I was looking at reflect.Value's `FieldByName`. It's a nit so since it's not convenient feel free to keep to whatever you like. I don't mind.

**Comment by wenovus** [comment]
Sorry, another comment is we should mark this flag as "only tested for marshalling JSON". So note to the user that unmarshalling doesn't yet work (at least not until we add tests)

**Comment by hansthienpondt** [comment]
> Sorry, another comment is we should mark this flag as "only tested for marshalling JSON". So note to the user that unmarshalling doesn't yet work (at least not until we add tests)

Could you point me to the test cases for Unmarshaling JSON? I've only come across https://github.com/openconfig/ygot/blob/c15884d06738b435020d84e167f7143627bf19ea/ytypes/unmarshal_test.go#L23
Is this where a test should be defined?

In the Unmarshaling case, I'd expect this (an empty value) to be a use/test-case anyhow right? Whether this is a presence container or not, is not known yet at that point?
PS: the testing I've done so far, it looks like (un)marshalling works fine for presence containers, but obviously a test provides 100% guarantee

**Comment by wenovus** [comment]
There is also [TestUnmarshalContainer](https://github.com/openconfig/ygot/search?q=TestUnmarshalContainer), the tests are split by node type. That sounds good, if there is no extra implementation involved sounds good to me to just add the unmarshal tests in this PR and then we can say for this flag that "it's supported for JSON marshalling and unmarshalling". In particular this doesn't work with the `populate_defaults` flag, but as long as we mention that it's only supported for these tested operations this sounds good to me to add.

File: ygot/render_test.go (hunk 4 above)
Comment (hansthienpondt) [reply ygot/render_test.go]: done!

**Comment by hansthienpondt** [comment]
> There is also [TestUnmarshalContainer](https://github.com/openconfig/ygot/search?q=TestUnmarshalContainer), the tests are split by node type. That sounds good, if there is no extra implementation involved sounds good to me to just add the unmarshal tests in this PR and then we can say for this flag that "it's supported for JSON marshalling and unmarshalling". In particular this doesn't work with the `populate_defaults` flag, but as long as we mention that it's only supported for these tested operations this sounds good to me to add.

Done, could you check whether the test's I've added cover enough ground?

File: ygot/render_test.go [hunk 5]
```diff
@@ -3042,12 +3101,120 @@ func TestConstructJSON(t *testing.T) {
...
+			System: &ucExampleSystem{
+				SshServer: nil,
+			},
+		},
+		wantIETF: map[string]interface{}{
+			"bgp": map[string]interface{}{},
+		},
+		wantInternal: map[string]interface{}{
+			"bgp": map[string]interface{}{},
+		},
```
Comment (wenovus) [inline ygot/render_test.go:3217]: Add another test case with SshServer: &ucExampleSystemSshServer{}?

**Review by wenovus — COMMENTED** [review]
Thanks looks pretty good to me, just a couple minor comments.

**Review by wenovus — COMMENTED** [review]
```go
			if diff := pretty.Compare(gotietf, tt.wantIETF); diff != "" {
				t.Errorf("ConstructIETFJSON(%v): did not get expected output, diff(-got,+want):\n%v", tt.in, diff)
			}
```
For this block in render_test.go I'm wondering if we should convert to using `cmp.Diff`, which is the standard comparison tool which by default differentiates between the empty type and a nil value for comparing `map[string]interface{}`.

File: ytypes/container_test.go [hunk 6]
```diff
@@ -373,14 +394,42 @@ func TestUnmarshalContainer(t *testing.T) {
...
+			parent: &ParentContainerStruct{ContainerField: &ContainerStruct{ConfigLeaf1Field: ygot.Int32(1), StateLeaf1Field: ygot.Int32(2), Container2Field: nil}},
+			json:   `{ "container-field": { "leaf2-field": 43, "config": { "leaf1-field": 41 } , "state": { "leaf1-field": 42 } , "container2-field": { "leaf3-field": 44 } } }`,
+			want:   &ParentContainerStruct{ContainerField: &ContainerStruct{ConfigLeaf1Field: ygot.Int32(41), StateLeaf1Field: ygot.Int32(42), Leaf2Field: ygot.Int32(43), Container2Field: &Container2Struct{Leaf3Field: ygot.Int32(44)}}},
+		},
+		{
+			desc:   "success overwriting existing fields with empty presence container",
+			schema: containerSchema,
+			parent: &ParentContainerStruct{ContainerField: &ContainerStruct{ConfigLeaf1Field: ygot.Int32(1), StateLeaf1Field: ygot.Int32(2), Container2Field: nil}},
+			json:   `{ "container-field": { "leaf2-field": 43, "config": { "leaf1-field": 41 } , "state": { "leaf1-field": 42 } , "container2-field": {  } } }`,
+			want:   &ParentContainerStruct{ContainerField: &ContainerStruct{ConfigLeaf1Field: ygot.Int32(41), StateLeaf1Field: ygot.Int32(42), Leaf2Field: ygot.Int32(43), Container2Field: &Container2Struct{}}},
```
[the remark this answers is not in the document]
Comment (hansthienpondt) [reply ytypes/container_test.go]: Sorry, my bad, a copy-paste error and I overlooked the test-cases. Indeed `Container2Field` should have gotten a non-nil value! fixed now. 

Considering the semantics, I believe you're right and it's indeed a no-op. I assume "replace" behaviour can be achieved using Merge (MergeOverwriteExistingFields) ?

File: ygot/render_test.go (hunk 5 above)
Comment (hansthienpondt) [reply ygot/render_test.go]: added the empty struct in a test case for `SshServer: &ucExampleSystemSshServer{}`, also re-ordered the test cases a bit, so it's uniform testing empty value, nil value and with data in tree.

**Comment by hansthienpondt** [comment]
> ```go
> 			if diff := pretty.Compare(gotietf, tt.wantIETF); diff != "" {
> 				t.Errorf("ConstructIETFJSON(%v): did not get expected output, diff(-got,+want):\n%v", tt.in, diff)
> 			}
> ```
> 
> For this block in render_test.go I'm wondering if we should convert to using `cmp.Diff`, which is the standard comparison tool which by default differentiates between the empty type and a nil value for comparing `map[string]interface{}`.

We could, but existing test-cases would break (not functionally though):
```
--- FAIL: TestConstructJSON (0.01s)
    --- FAIL: TestConstructJSON/multi-element_render_ConstructIETFJSON (0.00s)
    --- FAIL: TestConstructJSON/json_test_with_complex_children_ConstructIETFJSON (0.00s)
    --- FAIL: TestConstructJSON/json_test_with_complex_children_with_PrependModuleNameIdentityref=true_ConstructIETFJSON (0.00s)
    --- FAIL: TestConstructJSON/device_example_#1_ConstructIETFJSON (0.00s)
    --- FAIL: TestConstructJSON/device_example_#2_ConstructIETFJSON (0.00s)
    --- FAIL: TestConstructJSON/union_example_-_string_ConstructIETFJSON (0.00s)
    --- FAIL: TestConstructJSON/binary_example_ConstructIETFJSON (0.00s)
    --- FAIL: TestConstructJSON/uncompressed_device_example_with_presence_containers_with_data_in_tree_ConstructIETFJSON (0.00s)
```
for example:
```
did not get expected output, diff(-got,+want):
        -               "updates":      []interface{}{string("AQID"), string("AQIDBA==")},
        +               "updates":      []string{"AQID", "AQIDBA=="},

        -                                               "peer-as":          uint32(5413),
        +                                               "peer-as":          int(5413),

        -                               "as":        uint32(15169),
        +                               "as":        int(15169),

```

I have staged the changes of the test-cases, so if It would be your preference to do the conversion in this PR, I'll push it!

File: ytypes/container_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (wenovus) [reply ytypes/container_test.go]: Makes sense. MergeOverwriteExistingFields is not replace behaviour though. To get that you might just have to do it manually or via SetNode.

**Review by wenovus — APPROVED** [review]
Since it's existing tests that break this LGTM. If you have the staged changes let's look at them in a separate PR.
